src/lib/array.jl

```diff
@@ -214,19 +215,27 @@ _tryreverse(m, x) = x
 _tryreverse(m::typeof(map), x::Union{AbstractVector, Tuple}) = reverse(x)
 
 for (mapfunc,∇mapfunc) in [(:map,:∇map),(:pmap,:∇pmap)]
-  @eval function $∇mapfunc(cx, f, args...)
+  @eval function $∇mapfunc(cx, f::F, args...) where {F}
     ys_and_backs = $mapfunc((args...) -> _pullback(cx, f, args...), args...)
     if isempty(ys_and_backs)
       ys_and_backs, _ -> nothing
     else
-      ys, backs = unzip(ys_and_backs)
+      ys = map(first, ys_and_backs)
       ys, function (Δ)
         isnothing(Δ) && return nothing
-        # Apply pullbacks in reverse order. Needed for correctness if `f` is stateful.
-        Δf_and_args_zipped = $mapfunc((f, δ) -> f(δ), _tryreverse($mapfunc, backs, Δ)...)
-        Δf_and_args = unzip(_tryreverse($mapfunc, Δf_and_args_zipped))
-        Δf = reduce(accum, Δf_and_args[1])
-        (Δf, Δf_and_args[2:end]...)
+        if Base.issingletontype(F) && length(args) == 1
```

Comment:
Variants of the weird `map` example from existing tests. Tests pass, but at the REPL it would fail with this PR; some other variants are fine: 
```
julia> s = 0;

julia> function f(x)
            global s += x
          end;

julia> gradient(x -> sum(map(f, x)), 1:10)
([1, 2, 3, 4, 5, 6, 7, 8, 9, 10],)  # this PR
([10, 9, 8, 7, 6, 5, 4, 3, 2, 1],)  # tagged version, test is == (10:-1:1,)

julia> fieldnames(typeof(f))
()

julia> Base.issingletontype(typeof(f))
true

julia> S = Ref(0)  # mutable, global scope
Base.RefValue{Int64}(0)

julia> function F(x)
            S[] += x
          end;

julia> fieldnames(typeof(F))
()

julia> gradient(x -> sum(map(F, x)), 1:10)
([1, 2, 3, 4, 5, 6, 7, 8, 9, 10],)

julia> let
       S2 = Ref(0)  # mutable, local scope
       function F2(x)
           S2[] += x
       end
       gradient(x -> sum(map(F2, x)), 1:10)
       end
([10, 9, 8, 7, 6, 5, 4, 3, 2, 1],)

julia> let
       s2 = 0  # immutable, local scope
       function f2(x)
           s2 += x
       end
       gradient(x -> sum(map(f2, x)), 1:10)
       end
([10, 9, 8, 7, 6, 5, 4, 3, 2, 1],)
```

Reply:
yeah, probably we could afford a regression in those very corner cases. I don't know why things in the Main module act so differently. Maybe we could exclude methods defined in Main from the fast path? (I don't like this idea too much though)

Reply:
Yea there might be a way to test for this, but `parentmodule(f2)` is also `Main` so I don't think that's quite right.

But the example seems pretty contrived. We're concerned about getting gradients right, which sometimes depends on evaluation order. This is a naked test of evaluation order, as a proxy for the real concern. Perhaps it's not a close enough one, and we should test slightly more realistic things.

More generally I'm not sure Julia makes guarantees about the order of evaluation of `map` or `broadcast`. Especially for `broadcast`. I'm not too sure what that means for AD yet.

Reply:
I forget about Zygote's support for rebinging of globals.
It is like absolutely not the kind of mutation support that any one ever wants.
Reading globals used while taking derivatives of  yes,
taking derivatives with respect to globals: maybe, for interactive use in REPL, maaybe.
rebinding local variables, absolutely yes, so can have `for` loops.

but those globals getting rebound? I don't see it.
Almost, no good julia-code does that in the first place, especially not nice mathematicalish code that is worth ADing.
Even for tracking something into a global vector you want to mutate the value not rebind it.
It could be that we do screw that up though with reordering.

Reply:
ok, then I think I'm fine with merging this PR as it is